src/models/kinflated_gamma_poisson_model.py

```diff
@@ -0,0 +1,602 @@
+        self._kmax = kmax
+        self._debug = debug
+        self._fit_computed = False
+        self._extrapolation_multiplier = extrapolation_multiplier
+        self._extrapolation_value = extrapolation_multiplier * data[0].impressions[0]
+        if data[0].spends:
+            self._cpi = data[0].spends[0] / data[0].impressions[0]
+        else:
+            self._cpi = None
+
+    def _chi2_distance(
+        self,
+        h: npt.ArrayLike,
+        I: float,
+        N: float,
+        dist: KInflatedGammaPoissonDistribution,
+    ) -> float:
+        """Returns distance between actual and expected histograms.
+
+        Computes the metric
+
+            chi2(h, hbar) = \sum_i (h[i] - hbar[i])^2 / hbar[i]
+
+        where
+            hbar is the histogram that would be expected given parameters
+               I, N, dist
+
+        The first part of this objective formula is the chi-squared statistic
+        used for determining whether two sets of categorical samples are drawn
+        from the same underlying distribution.  Therefore, if the value returned
+        by this function is relatively low (say < 20), then there is reason to
+        believe that the Gamma-Poisson distribution adequately fits the data.
+
+        Args:
+          h: Actual frequency histogram.  h[i] is the number of people
+             reached exactly i+1 times, except for h[-1] which is the
+             number of people reached len(h) or more times.
+          I: Total number of impressions purchased.
+          N: Total number of unique users.
+          dist:   A KInflatedGammaPoisson distribution.
+        Returns:
+          chi^2 distance between the actual and expected histograms, plus
+          a term for weighting the difference in 1+ reach.
+        """
+        # Estimate total number of potential impressions
+        Imax = N * dist.expected_value()
+        if Imax <= I:
+            return np.sum(np.array(h) ** 2)
+
+        freqs = N * dist.kreach(np.arange(1, len(h)), I / Imax)
+        kplus = N * dist.kplusreach(len(h), I / Imax)
+        hbar = np.array(list(freqs) + [kplus])
+
+        obj = np.sum((hbar - h) ** 2 / (hbar + 1e-6))
+
+        if np.isnan(obj):
+            logging.vlog(2, f"alpha {alpha} beta {beta} N {N} Imax {Imax}")
+            logging.vlog(2, f"h    {h}")
+            logging.vlog(2, f"hbar {hbar}")
+            raise RuntimeError("Invalid value of objective function")
+
+        return obj
+
+    def _exponential_poisson_reach(self, I: float, N: float, beta: float) -> float:
+        """Estimates reach in an exponential poisson model.
+
+        Estimates reach when I impressions are purchased from a population
+        of N users where the impression count of each user is distributed
+        according to the exponential poisson distribution with parameter beta.
+
+        Under the exponential poisson model, the probability that a
+        user will have n impressions is (1 - p) * p^{n-1}, where p =
```

Comment:
is f(n) = ...
since you mentioned f(n) below

Reply:
Good catch.  Thanks.